api/search.go

```diff
@@ -1,28 +1,632 @@
+				continue
+			}
+			err = json.Unmarshal(temp, &meiliCourses)
+			if err != nil { // shouldn't happen
+				continue
+			}
+
+			for _, meiliCourse := range meiliCourses {
+				course, err := daoWrapper.CoursesDao.GetCourseBySlugYearAndTerm(c, meiliCourse.Slug, meiliCourse.TeachingTerm, meiliCourse.Year)
+				if err == nil && user.IsEligibleToSearchForCourse(course) {
+					res.Hits = append(res.Hits, meiliCourse)
+				}
+
+				if len(res.Hits) >= int(limit) {
+					break
+				}
+			}
+			response.Results[i] = res
+		case "SUBTITLES":
+			hits := res.Hits
+			res.Hits = []any{}
+			response.Results[i] = meilisearch.SearchResponse{}
+
+			var meiliSubtitles []SearchSubtitlesDTO
+			temp, err := json.Marshal(hits)
+			if err != nil { // shouldn't happen
+				continue
+			}
+			err = json.Unmarshal(temp, &meiliSubtitles)
+			if err != nil { // shouldn't happen
+				continue
+			}
+
+			for _, meiliSubtitle := range meiliSubtitles {
+				stream, err := daoWrapper.StreamsDao.GetStreamByID(c, strconv.FormatUint(uint64(meiliSubtitle.StreamID), 10))
+				if err != nil {
+					continue
+				}
+				course, err := daoWrapper.CoursesDao.GetCourseById(c, stream.CourseID)
+				if err != nil {
+					continue
+				}
+
+				meiliSubtitle.StreamName = stream.Name
+				meiliSubtitle.StreamStartTime = stream.Start
+				meiliSubtitle.StreamEndTime = stream.End
+				meiliSubtitle.CourseSlug = course.Slug
+				meiliSubtitle.CourseName = course.Name
+				meiliSubtitle.CourseYear = course.Year
+				meiliSubtitle.CourseTeachingTerm = course.TeachingTerm
+				if userEligibleToSeeResultsOfHiddenCourse(course) && (!stream.Private || user.IsAdminOfCourse(course)) {
+					res.Hits = append(res.Hits, meiliSubtitle)
+				}
+
+				if len(res.Hits) >= int(limit) {
+					break
+				}
+			}
+			response.Results[i] = res
+		default:
+			continue
+		}
+	}
+}
+
+// meilisearch filters
+
+// meiliSubtitleFilter returns a filter conforming to MeiliSearch filter format that can be used for filtering subtitles
+//
+// Checking eligibility to search in each course in courses is the caller's responsibility
+func meiliSubtitleFilter(user *model.User, courses []model.Course) string {
+	if len(courses) == 0 {
+		return ""
+	}
+
+	var streamIDs []uint
+	for _, course := range courses {
+		admin := user.IsAdminOfCourse(course)
+		for _, stream := range course.Streams {
+			if !stream.Private || admin {
+				streamIDs = append(streamIDs, stream.ID)
+			}
+		}
+	}
+	return fmt.Sprintf("streamID IN %s", uintSliceToString(streamIDs))
+}
+
+// meiliStreamFilter returns a filter conforming to MeiliSearch filter format that can be used for filtering streams
+//
+// Checking eligibility to search for courses and validation of model.Semester format is the caller's responsibility
+func meiliStreamFilter(c *gin.Context, user *model.User, semester model.Semester, courses []model.Course) string {
+	if courses != nil {
+		return fmt.Sprintf("courseID IN %s", courseSliceToString(courses))
+	}
```

Comment:
This ignores the case, where the user is not allowed to see private streams. Even if this does not lead to data leaks because of `checkAndFillResponse`, this should be dealt with.

Reply:
Commit 8295e423d7108f8e8b3393aa92c70c102ba26524 solves that issue.